File: plugin-dotnet-agent/src/test/kotlin/jetbrains/buildServer/dotnet/test/StringExtensions.kt
package jetbrains.buildServer.dotnet.test

import java.io.File

internal object StringExtensions {
    private const val WINDOWS_SEPARATOR = '\\'
    private const val UNIX_SEPARATOR = '/'
    private const val WINDOWS_DRIVE_C = 'C'

    // Handles string as path and normalizes it in an OS-agnostic way.
    fun String.toPlatformPath(drive: Char = WINDOWS_DRIVE_C) = when {
        File.separatorChar == UNIX_SEPARATOR -> this.replace(WINDOWS_SEPARATOR, UNIX_SEPARATOR)
        else -> this.replace(UNIX_SEPARATOR, WINDOWS_SEPARATOR).let { windowsPath ->
            if (this.startsWith(UNIX_SEPARATOR.toString())) "$drive:$windowsPath" else windowsPath
        }
    }
}

File: plugin-dotnet-agent/src/test/csharp/TeamCity.Dotnet.TestSuppressor/TeamCity.Dotnet.TestSuppressor.UnitTests/StringExtensions.cs
namespace TeamCity.Dotnet.TestSuppressor.UnitTests;

internal static class StringExtensions
{
    private static class PathSeparator
    {
        public const char Windows = '\\';
        public const char Unix = '/';
    }

    private static class WindowsDrive
    {
        public const char C = 'C';
    }

    /// <summary>
    /// Handles string as path and normalize it in OS agnostic way
    /// </summary>
    public static string ToPlatformPath(this string path, char drive = WindowsDrive.C)
    {
        if (Path.DirectorySeparatorChar == PathSeparator.Unix)
        {
            return path.Replace(PathSeparator.Windows, PathSeparator.Unix);
        }
        
        var windowsPath = path.Replace(PathSeparator.Unix, PathSeparator.Windows);
        return path.StartsWith(PathSeparator.Unix.ToString()) ? $"{drive}:{windowsPath}" : windowsPath;
    }
}
